File: bindings/openai_api.py
import threading
from typing import Any, Iterable, List, Union
import os, sys, signal, queue, time
from http.server import HTTPServer, BaseHTTPRequestHandler
from chatllm import LibChatLLM, ChatLLM, LLMChatInput, LLMChatDone, LLMChatChunk, ChatLLMStreamer
import json

MODEL = 'chatllm-variant'

class HttpResponder:
    def __init__(self, req: BaseHTTPRequestHandler, id: str, timestamp: int, model: str) -> None:
        self.req = req
        self.timestamp = timestamp
        self.model = model
        self.id = id

    def recv_chunk(self, chunk: str) -> bool:
        return True

    def done(self) -> None:
        pass

    def send_str(self, s: str) -> bool:
        self.req.wfile.write(s.encode('utf-8'))
        return True

class ChatCompletionNonStreamResponder(HttpResponder):
    def __init__(self, req: BaseHTTPRequestHandler, id: str, timestamp: int, model: str) -> None:
        super().__init__(req, id, timestamp, model)
        self.acc = ''

    def recv_chunk(self, content: str) -> bool:
        self.acc = self.acc + content
        return True

    def done(self) -> None:
        rsp = {
                "id": self.id,
                "object": "chat.completion",
                "created": self.timestamp,
                "model": self.model,
                "system_fingerprint": "fp_xxx",
                "choices": [{
                    "index": 0,
                    "delta": {
                        "role": "assistant",
                        "content": self.acc,
                    },
                    "logprobs": None,
                    "finish_reason" : "stop"
                }],
                'usage': {
                    "prompt_tokens": 1,
                    "completion_tokens": 1,
                    "total_tokens": 1
                }
            }
        self.send_str(json.dumps(rsp) + '\n\n')

class ChatCompletionStreamResponder(HttpResponder):
    def __init__(self, req: BaseHTTPRequestHandler, id: str, timestamp: int, model: str) -> None:
        super().__init__(req, id, timestamp, model)

    def recv_chunk(self, content: str) -> bool:
        rsp = {
                "id": self.id,
                "object": "chat.completion.chunk",
                "created": self.timestamp,
                "model": self.model,
                "system_fingerprint": "fp_xxx",
                "choices": [{
                    "index": 0,
                    "delta": {
                        "role": "assistant",
                        "content": content,
                    },
                    "logprobs": None,
                    "finish_reason" : None
                }]
            }
        self.send_str(json.dumps(rsp) + '\n\n')
        return True

    def done(self) -> None:
        rsp = {
                "id": self.id,
                "object": "chat.completion",
                "created": self.timestamp,
                "model": self.model,
                "system_fingerprint": "fp_xxx",
                "choices": [{
                    "index": 0,
                    "delta": {
                        "role": "assistant",
                        "content": '',
                    },
                    "logprobs": None,
                    "finish_reason" : "stop"
                }],
                'usage': {
                    "prompt_tokens": 1,
                    "completion_tokens": 1,
                    "total_tokens": 1
                }
            }
        self.send_str(json.dumps(rsp) + '\n\n')

class LegacyCompletionNonStreamResponder(HttpResponder):
    def __init__(self, req: BaseHTTPRequestHandler, id: str, timestamp: int, model: str) -> None:
        super().__init__(req, id, timestamp, model)
        self.acc = ''

    def recv_chunk(self, content: str) -> bool:
        self.acc = self.acc + content
        return True

    def done(self) -> None:
        rsp = {
            "id": self.id,
            "object": "text_completion",
            "created": self.timestamp,
            "model": self.model,
            "system_fingerprint": "fp_xxx",
            "choices": [
                {
                "text": self.acc,
                "index": 0,
                "logprobs": None,
                "finish_reason": "length"
                }
            ],
            "usage": {
                "prompt_tokens": 1,
                "completion_tokens": 1,
                "total_tokens": 1
            }
        }
        self.send_str(json.dumps(rsp) + '\n')

class LegacyCompletionStreamResponder(HttpResponder):
    def __init__(self, req: BaseHTTPRequestHandler, id: str, timestamp: int, model: str) -> None:
        super().__init__(req, id, timestamp, model)
        self.first_chunk = True

    def recv_chunk(self, content: str) -> bool:
        rsp = {
                "id": self.id,
                "object": "chat.completion.chunk",
                "created": self.timestamp,
                "choices": [
                {
                    "text": content,
                    "index": 0,
                    "logprobs": None,
                    #"finish_reason": None
                }
                ],
                "model": self.model,
                "system_fingerprint": "fp_xxx"
            }

        self.send_str('data: ' + json.dumps(rsp) + '\n\n')
        return True

    def done(self) -> None:
        self.send_str('data: [DONE]\n')

class SessionManager:
    def __init__(self) -> None:
        self._id = 1

    def make_id(self) -> str:
        self._id += 1
        return '_chatllm_' + str(self._id)

session_man: SessionManager = SessionManager()
llm_streamer: ChatLLMStreamer = None

class HttpHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(404, 'POST')

    def do_POST(self):
        print(self.path)
        args = self.rfile.read(int(self.headers['content-length'])).decode('utf-8')
        try:
            obj = json.loads(args)
            print(obj)
        except:
            self.send_error(404, 'BAD REQ')
            return

        if (self.path != '/v1/completions') and (self.path != '/v1/chat/completions') and \
           (self.path != '/completions') and (self.path != '/chat/completions'):
            self.send_error(404, 'NOT FOUND')
            return

        self.send_response(200)

        id = session_man.make_id()
        timestamp = int(time.time())
        prompt = ''
        stream = False
        if 'stream' in obj:
            stream = obj['stream']
        if 'messages' in obj:
            for x in obj['messages']:
                if x['role'] == 'user':
                    prompt = x['content']

            responder_cls = ChatCompletionStreamResponder if stream else ChatCompletionNonStreamResponder

            self.send_header('Content-type', 'application/json')
        else:
            prompt = obj['prompt']
            responder_cls = LegacyCompletionStreamResponder if stream else LegacyCompletionNonStreamResponder

            if stream:
                self.send_header('Content-type', 'text/event-stream')
            else:
                self.send_header('Content-type', 'application/json')

        self.end_headers()

        responder = responder_cls(self, id, timestamp, MODEL)
        for x in llm_streamer.chat(prompt):
            #print(x)
            try:
                responder.recv_chunk(x)
            except:
                llm_streamer.abort()

        responder.done()

def handler(signal_received, frame):
    sys.exit(0)

if __name__ == '__main__':
    signal.signal(signal.SIGINT, handler)

    args = sys.argv[1:]
    llm = ChatLLM(LibChatLLM(), args, False)
    llm_streamer = ChatLLMStreamer(llm)
    httpd = HTTPServer(('localhost', 3000), HttpHandler)
    httpd.serve_forever()

File: bindings/openai_api.ts
var worker = new Worker('chatllm.ts');

worker.postMessage({id : '#start', argv: Bun.argv.slice(2)});

const model = 'chatllm-variant';

function make_chat_completion_chunk_obj(id: string, content: string, timestamp: number, model: string) {
    return {
        "id": id,
        "object": "chat.completion.chunk",
        "created": timestamp,
        "model": model,
        "system_fingerprint": "fp_xxx",
        "choices": [{
          "index": 0,
          "delta": {
            "role": "assistant",
            "content": content,
          },
          "logprobs": null,
          "finish_reason" : null
        }]
      };
}

function make_chat_completion_obj(id: string, timestamp: number, model: string, content: string = '') {
    let obj = make_chat_completion_chunk_obj(id, content, timestamp, model);
    obj.object = "chat.completion";
    obj.choices[0].finish_reason = "stop";
    obj.usage = {
        "prompt_tokens": 1,
        "completion_tokens": 1,
        "total_tokens": 1
    }
    return obj;
}

class ChatCompletionSession {
    controller: ReadableStreamDefaultController;
    model: string;
    created: number;
    streaming: boolean;
    acc: string;

    constructor(controller: ReadableStreamDefaultController, model: string, streaming: boolean) {
        this.controller = controller;
        this.model = model;
        this.created = Date.now();
        this.streaming = streaming;
        this.acc = '';
    }

    receive_chunk(id: string, chunk: string) {
        if (this.streaming) {
            this.controller.enqueue('data: ' + JSON.stringify(make_chat_completion_chunk_obj(id, chunk, this.created, this.model)) + '\n\n');
        } else {
            this.acc = this.acc + chunk;
        }
    }

    complete(id: string) {
        if (this.streaming) {
            this.controller.enqueue('data: ' + JSON.stringify(make_chat_completion_obj(id, this.created, this.model)) + '\n\n');
        } else {
            this.controller.enqueue('data: ' + JSON.stringify(make_chat_completion_obj(id, this.created, this.model, this.acc)) + '\n\n');
        }
    }
}

function make_legacy_chat_complete_chunk_obj(id: string, timestamp: number, model: string, content: string) {
    return {
        "id": id,
        "object": "text_completion",
        "created": timestamp,
        "choices": [
          {
            "text": content,
            "index": 0,
            "logprobs": null,
            "finish_reason": null
          }
        ],
        "model": model,
        "system_fingerprint": "fp_xxx"
      };
}

function make_legacy_chat_complete_obj(id: string, timestamp: number, model: string, content: string) {
    return {
        "id": id,
        "object": "text_completion",
        "created": timestamp,
        "model": model,
        "system_fingerprint": "fp_xxx",
        "choices": [
            {
            "text": content,
            "index": 0,
            "logprobs": null,
            "finish_reason": "length"
            }
        ],
        "usage": {
            "prompt_tokens": 1,
            "completion_tokens": 1,
            "total_tokens": 1
        }
    };
}

class LegacyChatCompletionSession {
    controller: ReadableStreamDefaultController;
    model: string;
    created: number;
    streaming: boolean;
    acc: string;
    first_chunk: boolean;

    constructor(controller: ReadableStreamDefaultController, model: string, streaming: boolean) {
        this.controller = controller;
        this.model = model;
        this.created = Date.now();
        this.streaming = streaming;
        this.acc = '';
        this.first_chunk = true;
    }

    receive_chunk(id: string, chunk: string) {
        if (this.streaming) {
            if (this.first_chunk) {
                this.first_chunk = false;
                this.controller.enqueue(JSON.stringify(make_legacy_chat_complete_chunk_obj(id, this.created, this.model, chunk)) + '\n');
            }
            this.controller.enqueue('data: ' + chunk + '\n');
        } else {
            this.acc = this.acc + chunk;
        }
    }

    complete(id: string) {
        if (this.streaming) {
            this.controller.enqueue('data: [DONE]\n');
        } else {
            this.controller.enqueue('data: ' + JSON.stringify(make_legacy_chat_complete_obj(id, this.created, this.model, this.acc)) + '\n\n');
        }
    }
}

var map_id2session: {[key: string]: ChatCompletionSession | LegacyChatCompletionSession | null} = {};
var id_counter = 1;

function create_new_id() {
    let id = 'chatllm-' + id_counter;
    id_counter += 1;
    return id;
}

function create_new_chat_complete(id: string, controller: ReadableStreamDefaultController, streaming: boolean) {
    map_id2session[id] = new ChatCompletionSession(controller, model, streaming);
}

function create_new_legacy_chat_complete(id: string, controller: ReadableStreamDefaultController, streaming: boolean) {
    map_id2session[id] = new LegacyChatCompletionSession(controller, model, streaming);
}

function cancel_chat_session(id: string) {
    delete map_id2session[id];
}

worker.onmessage = function(e) {
    let sess = map_id2session[e.data.id];

    if ((sess instanceof ChatCompletionSession) || (sess instanceof LegacyChatCompletionSession)) {
        if (e.data.type == 'chunk') {
            sess.receive_chunk(e.data.id, e.data.content);
        } else if (e.data.type == 'end') {
            sess.complete(e.data.id);
            sess.controller.close();
            delete map_id2session[e.data.id];
        }

        return;
    }

    console.log("unexpected session", sess);
}

async function chat_completion(req: Request, obj: any) {
    let prompt = '';
    for (let msg of obj.messages) {
        if (msg.role == 'user') prompt = msg.content;
    }
    if (prompt == '') return;

    let id = create_new_id();
    let streaming = obj.stream == true;

    let streamer = new ReadableStream({
        start(controller) {
            create_new_chat_complete(id, controller, streaming);
            worker.postMessage({id: id, user: prompt});
        },
        pull(controller) {
        },
        cancel() {
            cancel_chat_session(id);
        },
    });

    return new Response(streamer);
}

async function legacy_chat_completion(req: Request, obj: any) {
    let prompt = obj.prompt;
    let id = create_new_id();
    let streaming = obj.stream == true;

    let streamer = new ReadableStream({
        start(controller) {
            create_new_legacy_chat_complete(id, controller, streaming);
            worker.postMessage({id: id, user: prompt});
        },
        pull(controller) {
        },
        cancel() {
            cancel_chat_session(id);
        },
    });

    return new Response(streamer);
}

Bun.serve({
    port: 3000,
    async fetch(req: Request) {
        if (req.method != "POST") return new Response("Please use POST");
        if (req.url.endsWith("v1/chat/completions") || req.url.endsWith("v1/completions")) {
            let body = await (await req.blob()).text();
            let obj = JSON.parse(body);
            if (typeof(obj.prompt) == 'string') return await legacy_chat_completion(req, obj);
            return await chat_completion(req, obj);
        }
        return new Response("404");
    },
    error(error) {
        return new Response(`<pre>${error}\n${error.stack}</pre>`, {
            headers: {
                "Content-Type": "text/html",
            },
        });
    },
});